### PR TITLE
[5.4] Ability to remove middlewares for specific route

### PR DESCRIPTION
**Pull request by deleugpn**
This allows us to remove middleware from a very specific route without having to rely on RouteServiceProvider to create a whole new Router environment for a single route.

```
Route::get('/health', 'ClosuresController@health')->withoutMiddleware();
```

I decided not to put too much effort into it to get a feedback before moving with a more interesting approach where we can remove just a single middleware. The following snippet would be the next step.

```
Route::get('/health', 'ClosuresController@health')->withoutMiddleware(\App\Http\Middleware\VerifyCsrfToken::class);
```

### CODE REVIEW

**Comment by tomschlick** [comment]
This would be great. Would you see this applying to route groups that the route is nested in as well? Or would those still apply the middleware?

As someone who has hundreds of nested routes, sometimes I just want to remove the middleware for one route deep in that nest (ex. to allow a guest user to download/view information).

The workarounds are to specify the middleware on the controller or to duplicate the route structure for that one route... but those suck if you want a single source of truth for your routes & middleware.

**Comment by taylorotwell** [comment]
I think it's a cool feature. I would go ahead and add in the ability to remove only a subset of the assigned middleware by passing parameters.

**Comment by JosephSilber** [comment]
To be clear: this will obviously only remove the route middleware, not the global middleware.

Might be a little confusing.

**Comment by taylorotwell** [comment]
Hmm, yeah that is something to keep in mind.

**Comment by deleugpn** [comment]
I'm thinking about moving this to 5.5 because the `Route` doesn't know which middlewares it will have if you assign a middleware group to it. For instance, routes on `web.php` all receives the `web` middleware group. Only when the `Router` will decide the next step of the request, it translates `web` into `middlewareGroups` in the [Gather Route Middleware method](https://github.com/laravel/framework/blob/5.4/src/Illuminate/Routing/Router.php#L570). 

Basically we either need to change how `Route->gatherMiddlware()` works by returning two indexes like this

```
[
    'include' => 'web',
    'exclude' => \App\Http\Middleware\VerifyCsrfToken::class
]
```

Which is not pretty and have the downside of changing behavior or we need to translate group middlewares when withoutMiddleware is called, which duplicates code between `Router` and `Route`.

**Comment by deleugpn** [comment]
The names might need a little adjustment, but the change in core seems fairly small for the benefit. This will work without parameters (remove all route middlewares) as well as with parameters as pointed out in the first post.

```
Route::get('/health', 'ClosuresController@health')->withoutMiddleware(\App\Http\Middleware\VerifyCsrfToken::class);
```

**Comment by deleugpn** [comment]
@tomschlick I tested and it seems to work as you described.

```
Route::group(['middleware' => 'tenant'], function () {
	Route::get('/elb-health', 'ClosuresController@health')->name('health')->withoutMiddleware(\App\Http\Middleware\Https::class, \App\Http\Middleware\Tenant::class);
});
```

In this snippet, the route does not receive the middleware `tenant` declared in the group. Is this what you meant?

**Comment by tomschlick** [comment]
@deleugpn Yes! Thats exactly what I meant. If this gets merged I owe you a beer! 🍺

**Comment by laurencei** [comment]
Do you think it should allow for the names of the middleware as well? i.e.

```
Route::group(['middleware' => 'auth'], function () {
	Route::get('/dashboard', 'ClosuresController@dashboard')->withoutMiddleware(`auth`);
});
```

But that might be a bit harder...

**Comment by deleugpn** [comment]
Looks like this PR got stuck. I can't see @laurencei's comment anymore, but I was able to resolve the middleware's name.

**Comment by tomschlick** [comment]
Looks like travis CI is having issues as well. https://www.traviscistatus.com/incidents/93wfcv0ksqrz

**Comment by laurencei** [comment]
sorry @deleugpn - I think I removed it by accident.

Looks good

**Comment by deleugpn** [comment]
This is broken due to `array_diff` failing to convert Closures to strings.

**Comment by deleugpn** [comment]
@taylorotwell can you revise the new strategy?

**Comment by taylorotwell** [comment]
revise or review?

File: src/Illuminate/Routing/MiddlewareNameResolver.php [hunk 1]
```diff
@@ -14,13 +14,15 @@ class MiddlewareNameResolver
      * @param  array  $middlewareGroups
      * @return string|array
```
Comment (lucasmichot) [inline src/Illuminate/Routing/MiddlewareNameResolver.php]: Missing `@param` for `$disabledMiddlewares`

File: src/Illuminate/Routing/MiddlewareNameResolver.php [hunk 2]
```diff
@@ -51,17 +53,20 @@ public static function resolve($name, $map, $middlewareGroups)
      * @param  array  $middlewareGroups
      * @return array
```
Comment (lucasmichot) [inline src/Illuminate/Routing/MiddlewareNameResolver.php]: Missing `@param`

File: src/Illuminate/Routing/Route.php [hunk 3]
```diff
@@ -722,6 +732,29 @@ public function middleware($middleware = null)
     }
 
     /**
+     * Remove the given middlewares from the route. If no middleware is passed, all middlewares will be removed.
+     *
+     * @param null $middleware
```
Comment (lucasmichot) [inline src/Illuminate/Routing/Route.php]: @param mixed $middleware

File: src/Illuminate/Routing/Route.php [hunk 4]
```diff
@@ -722,6 +732,29 @@ public function middleware($middleware = null)
...
+     * @return $this
+     */
+    public function withoutMiddleware($middleware = null)
+    {
+        if (is_null($middleware)) {
```
Comment (lucasmichot) [inline src/Illuminate/Routing/Route.php]: Maybe use early return ?

```php
if ($middleware === null) {
    $this->action['middleware'] = [];

    return $this;
}

$middleware = is_array($middleware) ? $middleware : func_get_args();

$this->action['bypass'] = array_merge(
    (array) Arr::get($this->action, 'bypass', []), $middleware
);

return $this;
```

**Comment by taylorotwell** [comment]
I'm trying to think of other primary use cases for this. The example given isn't the best because you can already exclude certain URIs from CSRF verification in the middlewares $except property.

**Comment by deleugpn** [comment]
@taylorotwell in my case, I have a HTTPS middleware that I use in the `web` group middleware. But I need to exclude one web route from it. I didn't want to go to the RouteServiceProvider and make a whole new routing state just for one small route that will ignore HTTPS Middleware and respond for HTTP requests. Right now I'm using an if statement inside the Middleware to ignore the route, similar to the `$except` for CSRF.

@tomschlick might help us with this, too.

File: src/Illuminate/Routing/Route.php (hunk 4 above)
Comment (deleugpn) [inline src/Illuminate/Routing/Route.php]: I kept the `is_null` because the `middleware` method uses it and I was going for consistency. Do you think I should use $middlware === null anyway?

**Comment by tomschlick** [comment]
I have never even though of just checking the route in the middleware and excluding it via an if statement... That would completely solve my use case. I can't believe I didn't think of that earlier!

Also, does the `$except` property check exist on all middleware?! I thought it only existed on the CSRF middleware...

**Comment by deleugpn** [comment]
@tomschlick the `$except` is only for CSRF, I think. But since I needed it so much, I went to `Illuminate\Foundation\Http\Middleware\VerifyCsrfToken` and copied the `inExceptArray` method into my middleware and replicated the same behavior with the code.

File: src/Illuminate/Routing/Route.php (hunk 4 above)
Comment (lucasmichot) [inline src/Illuminate/Routing/Route.php]: `is_null` is fine too ;-)

**Comment by tomschlick** [comment]
I'd agree that a route declaration would be much clearer, but if it means that we are introducing a lot of complication into the routing logic then I'd be fine with the alternative. After all, the framework has to cater to the 95% use case.

**Comment by OwenMelbz** [comment]
@taylorotwell what about if the middleware is provided by a 3rd party so you can't ignore it at their level? is there any other solution to skip running it on a specific route?
